### ISSUE TITLE
triton 3.3 has a bit longer host overhead compare to 3.2

### ISSUE THREAD

**Issue by jianyizh** [intel/intel-xpu-backend-for-triton#3907]
### Describe the issue

I'm investigating a regression on torchbench model lennard_jones when updating pytorch from 2.6 to 2.7. It seems a regression from host overhead in triton. 
I fixed pytorch commit on 43372e70c278bb93d7696309b17221ad9e6df897, when triton commit is e98b6fcb8df5b44eb0d0addb6767c573d37ba024, it takes 1.9ms. when triton commit is updated to 0bcc8265e677e5321606a3311bf71470f14456a8, it takes 2.0ms. Note both case the kernels on device takes 0.2ms in total...

You can run this model by `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 bash inductor_xpu_test.sh torchbench amp_bf16 training performance xpu 0 static 1 0 lennard_jones` or  directly by `python benchmarks/dynamo/torchbench.py --performance --amp --amp-dtype=bfloat16 -dxpu -n10 --no-skip --dashboard --training --total-partitions 1 --partition-id 0 --only lennard_jones --backend=inductor --timeout=7200 --disable-cudagraphs --output=xxx.csv`

cm3leon_generate fp16 inference also have the same issue, trion 3.3 is ~880ms, 3.2 is ~800ms, both have kernel time ~300ms.

### Environment details

HW: spr+PVC1550

**Comment by jianyizh** [comment]
It's fine to put it low priority as the regression is only ~5% on this very small model. I create this issue for a reference

**Comment by alexbaden** [comment]
There are several improvements to kernel compilation and launching in upstream triton that didn’t make our 3.3 release. Plus we have oneaApi 2025.1 support coming soon. I suggest we remeasure on an updated PyTorch/triton pin and again with 2025.1.

**Comment by jianyizh** [comment]
Thanks, Let's check it in 2025.1.

**Comment by jianyizh** [comment]
@alexbaden Can you help make a pr to move the `os.getenv('TRITON_XPU_DUMP_SPIRV_KERNEL_ARGS', None)` into `__init__ ` https://github.com/intel/intel-xpu-backend-for-triton/blob/main/third_party/intel/backend/driver.py#L713? Although this is not related to the host regression, it still saves ~30ms e2e time on cm3leon_generate

**Comment by januszjah** [comment]
There are 1297 commits between 2 mentioned points. I did bisect and didn't find a single commit responsible for biggest change, it looks like if there is a change it's gradual increase in init time (I actually didn't see significant change), what seems natural when you add more and more features. If you think we need to improve host performance we can make specific issue for that.

Also it's worth mentioning I tested on pvc1100 instead of pvc1500.

**Comment by jianyizh** [comment]
PVC 1100 is slower than 1550, so host overhead maybe overlapped. I don't know what feature will increase the host launching time. Is it possible to compare the host time between cuda and xpu in triton (without sycl/cuda kernel submit)? We can then decide if we need to improve host performance.

**Comment by januszjah** [comment]
Well, you can compare nv vs xpu, just need the hardware. The easiest would be to compare xpu vs cpu but it seems not so insightful.

**Comment by jianyizh** [comment]
We can do the comparison if host issue is a blocker in our oob goal. About the regression itself, I don't think the current explanation is clear enough. Can you reproduce regression on cm3leon_generate? We cannot let host time "gradual increase" in future.

**Comment by januszjah** [comment]
@jianyizh I agree explanation is not clear enough and I'll provide more numbers.

**Comment by januszjah** [comment]
There was no change for lennard, but there is 5-10% change for cm3leon. I need to bisect these commits to find where change occured.

**Comment by januszjah** [comment]
I did a bisect but had to skip some commits due to compilation issues and the only info I got that's near to commit e3ac7d74445b9dde197532c0629065d259c7e4d0

**Comment by januszjah** [comment]
~~Performance dropped in https://github.com/intel/intel-xpu-backend-for-triton/commit/f936d127c3e12cb131cc2fec3552c905aeb86f02~~

**Comment by januszjah** [comment]
Commit responsible for a drop: https://github.com/intel/intel-xpu-backend-for-triton/commit/b70c7f7c1955d5c939595c5ad812b616ef0333ae
Reverting it helps.
However it makes sense that it has longer time on host startup time, because it has to compile arch_utils on the first use to query capabilities.
I'd mark it as *won't fix*, but I'd like to have input from author of the commit @whitneywhtsang

**Comment by whitneywhtsang** [comment]
We have modified the way we query support opencl extension, the current way is `sycl::opencl::has_extension`: https://github.com/intel/intel-xpu-backend-for-triton/blob/main/third_party/intel/backend/driver.c#L375
Is the host startup time better with the current way?
FYI @anmyachev

**Comment by januszjah** [comment]
I think the overhead comes from compiling arch_utils and current main still does it.

**Comment by jianyizh** [comment]
Does triton only use arch_utils or `has_opencl_extension` during init and compile? why e2e performance is affected after warmup?

**Comment by whitneywhtsang** [comment]
> Does triton only use arch_utils or `has_opencl_extension` during init and compile? why e2e performance is affected after warmup?

Yes, it does. It is unclear to me why it would affect e2e performance after warmup, theoretically it should not.

**Comment by whitneywhtsang** [comment]
@januszjah Can you please use prints to determine the number of times arch_utils get compiled and the number of times XPUBackend get created, when running the model?

**Comment by anmyachev** [comment]
> We have modified the way we query support opencl extension, the current way is `sycl::opencl::has_extension`: https://github.com/intel/intel-xpu-backend-for-triton/blob/main/third_party/intel/backend/driver.c#L375 Is the host startup time better with the current way? FYI [@anmyachev](https://github.com/anmyachev)

No it should be better, since `subprocess` module is no longer used.

**Comment by januszjah** [comment]
When I run: 
```
TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python3 benchmarks/dynamo/torchbench.py --performance --amp --amp-dtype=bfloat16 -dxpu -n10 --no-skip --dashboard --training --total-partitions 1 --partition-id 0 --only lennard_jones --backend=inductor --timeout=7200 --disable-cudagraphs --output=tmp.csv
```
Then it:
- Initialized XPUBackend up to 33 times per worker.
- Compiled arch_utils only once per worker, so ~7 times.

**Comment by januszjah** [comment]
It's not a problem with the cache & hash computation cost, I checked it.
I've changed the code to use precompiled .so for arch_utils and it helps, so it just a cost of compiling it.

Attached sample outputs from running pytorch benchmarks (as in bug description).

[runs.zip](https://github.com/user-attachments/files/23251322/runs.zip)

It was done on ubuntu 22.04 with DLE 2025.2   
filenames:
before & after - without and with mentioned commid
latest - latest main

**Comment by januszjah** [comment]
To sum up, the problem is that this case uses `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1` what is not-so-real scenario for arch_utils.so, that should be compiled at most once per Triton release - basically when source hash of .c file changes.

**Comment by jianyizh** [comment]
Thanks, I will check it later

**Comment by jianyizh** [comment]
close as I can get 800ms on cm3leon_generate without `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1`.  Thank you @januszjah for help
